nexus/reconfigurator/blippy/src/checks.rs

```diff
@@ -0,0 +1,524 @@
+    check_dataset_zpool_uniqueness(blippy);
+    check_datasets(blippy);
+}
+
+fn check_underlay_ips(blippy: &mut Blippy<'_>) {
+    let mut underlay_ips: BTreeMap<Ipv6Addr, &BlueprintZoneConfig> =
+        BTreeMap::new();
+    let mut inferred_sled_subnets_by_sled: BTreeMap<
+        SledUuid,
+        (Ipv6Subnet<SLED_PREFIX>, &BlueprintZoneConfig),
+    > = BTreeMap::new();
+    let mut inferred_sled_subnets_by_subnet: BTreeMap<
+        Ipv6Subnet<SLED_PREFIX>,
+        SledUuid,
+    > = BTreeMap::new();
+    let mut rack_dns_subnets: BTreeSet<DnsSubnet> = BTreeSet::new();
+
+    for (sled_id, zone) in blippy
+        .blueprint()
+        .all_omicron_zones(BlueprintZoneFilter::ShouldBeRunning)
+    {
+        let ip = zone.underlay_ip();
+
+        // There should be no duplicate underlay IPs.
+        if let Some(previous) = underlay_ips.insert(ip, zone) {
+            blippy.push_note(
+                Component::Sled(sled_id),
+                Severity::Fatal,
+                Kind::DuplicateUnderlayIp { zone1: previous, zone2: zone },
+            );
+        }
+
+        if zone.zone_type.is_internal_dns() {
+            // Internal DNS zones should have IPs coming from the reserved rack
+            // DNS subnets.
+            let subnet = DnsSubnet::from_addr(ip);
+            if rack_dns_subnets.is_empty() {
+                // The blueprint doesn't store the rack subnet explicitly, so we
+                // infer it based on the first internal DNS zone we see.
+                rack_dns_subnets.extend(subnet.rack_subnet().get_dns_subnets());
+            }
+            if !rack_dns_subnets.contains(&subnet) {
+                blippy.push_note(
+                    Component::Sled(sled_id),
+                    Severity::Fatal,
+                    Kind::InternalDnsZoneBadSubnet {
+                        zone,
+                        rack_dns_subnets: rack_dns_subnets.clone(),
+                    },
+                );
+            }
+        } else {
+            let subnet = Ipv6Subnet::new(ip);
+
+            // Any given subnet should be used by at most one sled.
+            match inferred_sled_subnets_by_subnet.entry(subnet) {
+                Entry::Vacant(slot) => {
+                    slot.insert(sled_id);
+                }
+                Entry::Occupied(prev) => {
+                    if *prev.get() != sled_id {
+                        blippy.push_note(
+                            Component::Sled(sled_id),
+                            Severity::Fatal,
+                            Kind::ConflictingSledSubnets {
+                                other_sled: *prev.get(),
+                                subnet,
+                            },
+                        );
+                    }
+                }
+            }
+
+            // Any given sled should have IPs within at most one subnet.
+            //
+            // The blueprint doesn't store sled subnets explicitly, so we can't
```

Comment:
Do you think the rack subnet is something we want to move into the blueprint eventually?

I think we do, mainly based on the "make everything in the blueprint explicit" POV.

Reply:
I think so, yeah. I don't think it's particularly urgent.